### ISSUE TITLE
Negative Sinkhorn divergences

### ISSUE THREAD

**Issue by johannbrehmer** [jeanfeydy/geomloss#25]
Hi,

First, thanks a lot for a great package that is very straightforward to use!

I am confused by some results I am getting where the Sinkhorn distance (with default parameters) is negative. From your website I understand that this shouldn't be the case.

Minimal example:
```
import numpy as np
import torch
from geomloss import SamplesLoss

x0 = torch.from_numpy(np.load("x0.npy"))
x1 = torch.from_numpy(np.load("x1.npy"))

sinkhorn = SamplesLoss("sinkhorn")

print(sinkhorn(x0, x1).item())
```
with the data files here:
[mwe.zip](https://github.com/jeanfeydy/geomloss/files/4135721/mwe.zip)

With geomloss v0.2.3 on the CPU this prints -0.022376153618097305.

Am I misunderstanding the Sinkhorn divergence, is there something anomalous in my data, or is this some numerical artefact?

Thanks a lot,
Johann

**Comment by jeanfeydy** [comment]
Hi @johannbrehmer ,

Thanks a lot for your detailed report!
A quick answer would be to tell you that this (unfortunate) negative output of the `SamplesLoss` layer is due to the default value of the `scaling` parameter that I agressively set to `0.5`. As discussed in [this tutorial](https://www.kernel-operations.io/geomloss/_auto_examples/sinkhorn_multiscale/plot_epsilon_scaling.html#sphx-glr-auto-examples-sinkhorn-multiscale-plot-epsilon-scaling-py), using a more conservative value of `0.7` or `0.9` should allow you to spend more time in the multiscale Sinkhorn loop (an annealing descent), and thus converge closer to the actual value of the (de-biased) Sinkhorn loss. In your case, something like `0.038 > 0`.

Nevertheless, since your example is one of the first that I encounter where changing the `scaling` parameter actually has a meaningful impact, I played around with it for a bit. Surprise: when we display it in 3D, we observe a few outliers lying far away from the unit sphere.

![counter_example](https://user-images.githubusercontent.com/20405678/73489423-da0ce180-43aa-11ea-8904-0b3c77c63afa.png)

If we remove these extreme points with a code along the lines of:

```python
x2 = x1.clone()
print(x2[(x1**2).sum(-1) > 1.5, :])
x2[(x1**2).sum(-1) > 1.5, :] = 0

print(sinkhorn(x0, x2).item())
```

We immediately retrieve a positive value even when `scaling = 0.5`. As far as I can tell, it thus looks like your example is a perfect worst-case scenario for the multiscale Sinkhorn loop implemented by GeomLoss. This is in line with the theoretical understanding of the algorithm that we're trying to develop with @bernhard-schmitzer : your configuration is pretty close to the "lone wolf" scenario discussed at Example 1, page 17 and Remark 9, page 21 of [this important reference paper](https://arxiv.org/pdf/1610.06519.pdf).

I'm pretty happy to see these theoretical considerations come to life in a real use-case :-)
If you have any other question, feel free to ask; otherwise, I'll let you close this issue.
Best regards,

Jean

P.S.: Out of curiosity, may I ask why you're interested in OT theory? You seem to work mostly on high-energy physics, and I'd be delighted to know how our codes may be of any use to you!

**Comment by johannbrehmer** [comment]
Hi Jean,

Thanks a lot for the fast and detailed answer! I'll have a closer look at the references you gave, but this makes sense to me.

I arrived at this worst-case scenario not randomly, but by training a generative model on minimizing the Sinkhorn divergence between generated samples and some training data. I had some anomalous results, now I hope that changing the scaling parameter might stabilize my training.

Thanks again,
Johann

PS: Indeed, my background is in high-energy physics, but I am now interested in probabilistic and generative ML models... hence OT. But I don't know much about the theory at all.
